lib/workers/testbot.js

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2018 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict'
+
+const _ = require('lodash')
+const Bluebird = require('bluebird')
+const Board = require('firmata')
+const drivelist = Bluebird.promisifyAll(require('drivelist'))
+const fs = Bluebird.promisifyAll(require('fs'))
+const mountutils = Bluebird.promisifyAll(require('mountutils'))
+const imageWrite = require('etcher-image-write')
+const visuals = require('resin-cli-visuals')
+
+const WRITE_DAC_REG = 0x00
+const ENABLE_VOUT_SW = 0x03
+const DISABLE_VOUT_SW = 0x04
+const ENABLE_VREG = 0x07
+const ENABLE_FAULTRST = 0x10
+const SD_RESET_ENABLE = 0x12
+const SD_RESET_DISABLE = 0x13
+
+const LED_PIN = 13
+const HW_SERIAL5 = 5
+const SD_MUX_SEL_PIN = 28
+const USB_MUX_SEL_PIN = 29
+
+const getDevInterface = async (devicePath) => {
+  const devIdLink = '/dev/disk/by-id/usb-Generic_Ultra_HS-SD_MMC_000008264001-0:0'
+
+  if (process.platform === 'linux') {
+    return fs.realpathAsync(devIdLink).catch((error) => {
+      throw new Error(`Could not find testbot. Reason: ${error}`)
+    })
+  }
+
+  if (_.isUndefined(devicePath)) {
+    throw new Error('For non-linux operating systems testbot dev interface has to be provided')
+  }
+
+  return devicePath
+}
+
+module.exports = class TestBotWorker {
+  constructor (title, deviceType, options = {}) {
+    this.title = title
+    this.deviceType = deviceType
+    this.options = options
+
+    this.board = new Board(options.devicePath)
+    this.board.serialConfig({
+      portId: HW_SERIAL5,
+      baud: 9600
+    })
+  }
+
+  sendCommand (command, a_ = 0, b_ = 0) {
+    this.board.serialWrite(HW_SERIAL5, [ command, a_, b_ ])
+  }
+
+  async connectSdToTarget (sw) {
+    this.sendCommand(SD_RESET_ENABLE)
+    await Bluebird.delay(10)
+    this.sendCommand(SD_RESET_DISABLE)
+
+    if (sw) {
+      this.board.digitalWrite(LED_PIN, this.board.LOW)
+      this.board.digitalWrite(SD_MUX_SEL_PIN, this.board.LOW)
+    } else {
+      this.board.digitalWrite(LED_PIN, this.board.HIGH)
+      this.board.digitalWrite(SD_MUX_SEL_PIN, this.board.HIGH)
+    }
+
+    return Bluebird.delay(2000)
+  }
+
+  powerTarget (sw) {
+    sw ? this.sendCommand(ENABLE_VOUT_SW) : this.sendCommand(DISABLE_VOUT_SW)
+  }
+
+  async flash (image) {
+    await this.connectSdToTarget(false)
```

Comment:
This looks a bit weird. Maybe we should have two functions like:

- `switchSDCardToTarget`
- `switchSDCardToHost`

?